sycl/doc/Assert.md

```diff
@@ -0,0 +1,138 @@
+```
+class event_error : public runtime_error {
+public:
+  event_error() = default;
+
+  event_error(const char *Msg, cl_int Err)
+      : event_error(string_class(Msg), Err) {}
+
+  event_error(const string_class &Msg, cl_int Err) : runtime_error(Msg, Err) {}
+};
+```
+
+Regardless of whether asynchronous exception handler is set or not, there's an
+action to be performed by SYCL Runtime. To achieve this, information about
+assert failure should be propagated from device-side to SYCL Runtime. This
+should be performed via calls to `clGetEventInfo` for OpenCL backend and
+`zeEventQueryStatus` for Level-Zero backend.
+
+## Terms
+
+ - Device-side Runtime - part of device-code, which is supplied by Device-side
+   Compiler.
+ - Low-level Runtime - the backend/runtime, behind DPCPP Runtime.
+ - Device-side Compiler - compiler which generates device-native bitcode based
+   on input SPIR-V image.
+ - Accessor metadata - parts of accessor representation at device-side: pointer,
+   ranges, offset.
+
+## How it works?
+
+For the time being, `assert(expr)` macro ends up in call to
+`__devicelib_assert_fail` function. This function is part of [Device library extension](doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+Device code already contains call to the function. Currently, a device-binary
+is always linked against fallback implementation.
+Device-side compiler/linker provides their implementation of `__devicelib_assert_fail`
+and prefer this implementation over fallback one.
+
+If Device-side Runtime supports `__devicelib_assert_fail` then Low-Level Runtime
+is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+at synchronization points.
+
+Refer to [OpenCL](doc/extensions/Assert/opencl.md) and [Level-Zero](doc/extensions/Assert/level-zero.md)
+extensions.
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a buffer
+based approach comes in place. The approach doesn't require any support from
+Device-side Runtime. Neither it does from Low-level Runtime.
+
+Within this approach, a dedicated assert buffer is allocated and implicit kernel
+argument is introduced. The argument is an accessor with `discard_read_write`
+or `discard_write` access mode. Accessor metadata is stored to program scope
+variable. This allows to refer to the accessor without modifying each and every
+user's function. Fallback implementation of `__devicelib_assert_fail` restores
+accessor metadata from program scope variable and writes assert information to
+the assert buffer. Atomic operations are used in order to not overwrite existing
+information.
+
+Storing and restoring of accessor metadata to/from program scope variable is
+performed with help of builtins. Implementations of these builtins are
+substituted by frontend.
```

Comment:
This section on the "dedicated buffer" approach seems too vague.  I think the design document should expand this description to give more details like:

* What are the compiler builtins and what do they do?
* More detail about how the kernel uses the hidden accessor.  Note that kernels cannot normally reference global variables, so the part about "program scope variable" seems suspect.
* What happens if multiple assertion failures write to the buffer.  Does the runtime raise one or many exceptions?  Do you print one message or all of them?  There could be many work items in flight, so printing 100's of error messages might not be a good thing.
* How do you avoid deadlock if a work item triggers an assertion and other work items are waiting on it?  A common example would be a work item that is participating in one of the "group" functions.  Normally, these functions block until all work items reach the "group" function.  What happens if one of the work items gets an assertion failure and never reaches that function?

Reply:
> Note that kernels cannot normally reference global variables, so the part about "program scope variable" seems suspect.

@gmlueck , could you, please, provide more details on this?

Within this approach, the global variable is not a SYCL variable but a LLVM IR/SPIR-V global variable. Specifically - per module global variable.

Reply:
> How do you avoid deadlock if a work item triggers an assertion and other work items are waiting on it? 

We can't stop kernel execution within this approach (no support for such sort of operation). Hence, if the deadlock is based on malformed data, it is unavoidable. E.g. nothing stops the user from typing in this code:
```
while(1) {
  assert(false && "This ain't work");
}
```

In this example the user won't get the kernel finished and the assert failure won't get reported.

Reply:
> > Note that kernels cannot normally reference global variables, so the part about "program scope variable" seems suspect.
> 
> @gmlueck , could you, please, provide more details on this?
> 
> Within this approach, the global variable is not a SYCL variable but a LLVM IR/SPIR-V global variable. Specifically - per module global variable.

Oh, I thought you planned to create a SYCL global variable.  I think I understand this part of the design now.  You will pass a parameter to the kernel that has the address of the "assert buffer" and then store this address in a kernel-side global variable.  The `__devicelib_assert_fail` function can then write the message into the buffer by reading the global variable.

I guess the big part of the design I don't understand now is how the device code aborts a kernel when an assertion happens and then signals the host, telling it that the kernel has aborted.

> ```
> while(1) {
>  assert(false && "This ain't work");
> }
> ```
>
> In this example the user won't get the kernel finished and the assert failure won't get reported.

This makes me think that there is no logic at all to abort a kernel after an assertion failure.  Is the following description accurate?

* A failing assert writes the message to the assertion buffer and then the `assert()` returns to its caller.
* The work item must continue to execute to completion after the failing assert.
* The kernel must complete even after a failing assert.
* After the DPC++ runtime sees that the kernel has completed, it checks the contents of the assertion buffer.
* If that buffer is not empty, the DPC++ runtime prints the message and reports an exception.

If that is the plan, I think we need to consider whether this is useful.  Applications do not expect a failed `assert()` call to return, and the code that follows often won't execute if the assertion condition is false.

If we plan native support for OpenCL and Level Zero, maybe that is enough and there is no need to have a fall-back implementation.

Reply:
> I guess the big part of the design I don't understand now is how the device code aborts a kernel when an assertion happens and then signals the host, telling it that the kernel has aborted.

When `__devicelib_assert_fail` implementation is available it's responsibility of Device-side Runtime to notify host. Aborting kernel execution isn't required.
When fallback implementation of `__devicelib_assert_fail` is used, DPCPP Runtime will read the assert buffer *after kernel finishes*.

>  Is the following description accurate?

Yes, it is for fallback implementation of `__devicelib_assert_fail`.

> If that is the plan, I think we need to consider whether this is useful. Applications do not expect a failed `assert()` call to return, and the code that follows often won't execute if the assertion condition is false.

There was suggestion from @JoeyGIntel to split naming of `assert` macro for supported by Device-side Runtime `__devicelib_assert_fail` and the fallback one.

> If we plan native support for OpenCL and Level Zero, maybe that is enough and there is no need to have a fall-back implementation.

That may not be enough. An example use-case is when Device-side Runtime/Compiler can't support for fully-featured assert due to H/W limitations.

Reply:
> When __devicelib_assert_fail implementation is available it's responsibility of Device-side Runtime to notify host. Aborting kernel execution isn't required.
> When fallback implementation of __devicelib_assert_fail is used, DPCPP Runtime will read the assert buffer after kernel finishes.

Why do you say that "aborting the kernel isn't necessary"?  My concern is that without some logic to abort the kernel, the kernel will often hang after an assert.  If it never finishes, the DPC++ runtime will never be notified, and then the user won't see the assert message.

> That may not be enough. An example use-case is when Device-side Runtime/Compiler can't support for fully-featured assert due to H/W limitations.

Do you know which of our H/W will have this limitation?